### PR TITLE
synchronize startup scan for all DBs

### PR DESCRIPTION
**Pull request by chappjc**
**Summary**

This PR changes how startup block sync/scan operates with regard to synchronization between the `StakeDatabase` ("stakedb"), `ChainDB` (postgresql), and `wiredDB` (sqlite).  Instead of sqlite and the stakedb, which is maintained by sqlite, and `ChainDB` running at their own pace, with loose communication between them (`ChainDB` required a pointer to the stakedb used by `wiredDB` to get winning tickets, etc.), now:

- `wiredDB` stays at the same height as `ChainDB`
- `ChainDB` is in charge of getting blocks.  They are provided to `wiredDB` via a RPC client wrapper, `BlockGetter`.  This limits getblock requests, and sets the pace for both `ChainDB` and `wiredDB`.
- `wiredDB` still maintains the stakedb, but as a result of the block RPC wrapper it waits for `ChainDB`.

_No DB rebuild is required._

**Terminology**
The master getter of blocks from dcrd is the "aux" DB (i.e. PostgreSQL right now).
The "base" DB (i.e. SQLite) connects blocks in a StakeDatabase instance.

**Changes**

- `BlockGate` and `[Master]BlockGetter`, for synchronization of backend rescans.
- `BlockGate` struct satisfies both `[Master]BlockGetter`
- `MasterBlockGetter` has `Update` methods to get a block via RPC and cache it
- `BlockGetter` has no `Update` methods and must use `WaitForHeight` or `WaitForHash`
- The `StakeDatabase` type now maintains a map of clients waiting for a certain height block, the clients (e.g. `ChainDB`) registered via `WaitForHeight`, and signals to the clients after each `ConnectBlock`
- During startup sync of dcrdata, use a `BlockGate` in place of a regular `rpcclient.Client`
- `ChainDB.SyncChainDBAsync` requires a `MasterBlockGetter`, which has `Update*` methods to request a block via RPC
- `wiredDB.SyncDBAsync` requires a `BlockGetter`, which must use `WaitForHeight`
- `ChainDB`'s sync thus uses `MasterBlockGetter.UpdateToBlock`, which gets the block via RPC and signals to all waiting on the block (i.e. `wireDB`), and then uses `StakeDatabase.WaitForHeight` to allow the stakedb to reach the height
- `wiredDB`'s sync uses `BlockGetter.WaitForHeight`, which gets the block when the `MasterBlockGetter` is ready, and only then does it run `StakeDatabase.ConnectBlock`, which signals to all waiting on the stakedb to reach a height (i.e. `ChainDB`)
- In lite mode, `wiredDB` is the "master" and uses a regular RPC client
-  recovery: Allow stakedb to catch up.  Add `fetchToHeight` to `BlockGate` to specify which calls to `WaitForHeight` do not wait, but immediately RPC. `fetchToHeight` means "start waiting at", so freely fetch all `<fetchToHeight` and wait above
- recovery: allow rewind of stakedb in addition to catch up: create `(*wiredDB).RewindStakeDB`
- allow CTRL+C to cleanly break channel receive from waitChan in wither dcrpg or dcrsqlite sync

Ancillary changes:

- `(*ChainDB).DeleteDuplicates` now removes duplicate transactions as well as vins and vouts
-  add methods to lock the internal `nodeMtx` of `StakeDatabase`
- Factor out DBHeights() function from wiredDB.resyncDB.
- Variable renaming:

sqliteDB --> baseDB
db --> auxDB
bestBlockHeight --> summaryHeight
bestStakeHeight --> stakeInfoHeight
bestNodeHeight --> stakeDBHeight

- Rename `*WithPoolValue` functions.
- do NOT call SyncDBAsync as a goroutine, as it gets the block ntfn channel, then launches a sync goroutine on its own.

### CODE REVIEW

**Comment by chappjc** [comment]
Addresses Issue https://github.com/dcrdata/dcrdata/issues/351

**Comment by chappjc** [comment]
Updated PR description.  Needs testing.  Tested w/ full mode, not lite mode.

File: db/dcrpg/sync.go [hunk 1]
```diff
@@ -46,10 +46,11 @@ func (db *ChainDB) SyncChainDBAsync(res chan dbtypes.SyncResult,
...
 // newIndexes to true. The quit channel is used to break the sync loop. For
 // example, closing the channel on SIGINT.
-func (db *ChainDB) SyncChainDB(client *rpcclient.Client, quit chan struct{},
+func (db *ChainDB) SyncChainDB(client rpcutils.MasterBlockGetter, quit chan struct{},
 	updateAllAddresses, updateAllVotes, newIndexes bool) (int64, error) {
 	// Get chain servers's best block
-	_, nodeHeight, err := client.GetBestBlock()
+	nodeHeight, err := client.NodeHeight()
+	//_, nodeHeight, err := client.GetBestBlock()
 	if err != nil {
```
Comment (chappjc) [inline db/dcrpg/sync.go]: remove

File: cmd/rebuilddb/rebuilddb.go [hunk 2]
```diff
@@ -107,7 +107,7 @@ func mainCore() int {
 	waitSync.Add(1)
 	//go sqliteDB.SyncDB(&waitSync, quit)
 	var height int64
-	height, err = sqliteDB.SyncDBWithPoolValue(&waitSync, quit)
+	height, err = sqliteDB.SyncDBWithPoolValue(&waitSync, quit, nil)
```
Comment (chappjc) [inline cmd/rebuilddb/rebuilddb.go]: build error.  Function teenaged

File: db/dcrsqlite/sync.go [hunk 3]
```diff
@@ -21,120 +22,68 @@ const (
...
 )
 
-func (db *wiredDB) resyncDB(quit chan struct{}) error {
-	// Get chain servers's best block
-	_, height, err := db.client.GetBestBlock()
-	if err != nil {
-		return fmt.Errorf("GetBestBlock failed: %v", err)
-	}
-
+func (db *wiredDB) DBHeights() (lowest int64, summaryHeight int64, stakeInfoHeight int64,
```
Comment (chappjc) [inline db/dcrsqlite/sync.go]: Docs

File: db/dcrsqlite/sync.go [hunk 4]
```diff
@@ -21,120 +22,68 @@ const (
...
+}
 
-	// Only save the block summaries we don't have in the DB yet
-	//blockSummaries := make([]apitypes.BlockDataBasic, 0, minBlocksToCheck+1)
+func (db *wiredDB) initWaitChan(waitChan chan chainhash.Hash) {
+	db.waitChan = waitChan
+}
 
-	for i = startHeight; i <= height; i++ {
+func (db *wiredDB) RewindStakeDB(toHeight int64, quit chan struct{}) (stakeDBHeight int64, err error) {
```
Comment (chappjc) [inline db/dcrsqlite/sync.go]: Docs

File: main.go [hunk 5]
```diff
@@ -182,92 +178,128 @@ func mainCore() error {
 		return fmt.Errorf("Unable to get block from node: %v", err)
 	}
 
+	var fetchToHeight = int64(math.MaxInt32)
 	if usePG {
```
Comment (chappjc) [inline main.go]: Comment that this is the default in lite mode meaning always RPC directly for blocks

File: main.go [hunk 6]
```diff
@@ -182,92 +178,128 @@ func mainCore() error {
...
-		stakeInfoHeight, err := sqliteDB.GetStakeInfoHeight()
+		// data from disk. The cache is updated on each block connect.
+		tpcSize := int(blocksBehind) + 20
+		log.Debugf("Setting ticket pool cache capacity to %d blocks", tpcSize)
+		err = baseDB.GetStakeDB().SetPoolCacheCapacity(tpcSize)
+		if err != nil {
+			return err
+		}
+
+		/*stakeInfoHeight, err := baseDB.GetStakeInfoHeight()
```
Comment (chappjc) [inline main.go]: Delete commented code here

File: rpcutils/blockgate.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+	NodeHeight() (int64, error)
+	BestBlockHeight() int64
+	BestBlockHash() (chainhash.Hash, int64, error)
+	BestBlock() (*dcrutil.Block, error)
+	Block(chainhash.Hash) (*dcrutil.Block, error)
+	WaitForHeight(int64) chan chainhash.Hash
+	WaitForHash(chainhash.Hash) chan int64
+}
+
+type MasterBlockGetter interface {
```
Comment (chappjc) [inline rpcutils/blockgate.go]: Docs

File: rpcutils/blockgate.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+// ensure BlockGate satisfies BlockGetter
+var _ BlockGetter = (*BlockGate)(nil)
+
+func NewBlockGate(client *rpcclient.Client, capacity int) *BlockGate {
```
Comment (chappjc) [inline rpcutils/blockgate.go]: Docs

File: rpcutils/blockgate.go [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+	block, hash, err := GetBlock(height, g.client)
+	if err != nil {
+		return nil, fmt.Errorf("GetBlock (%d) failed: %v", height, err)
+	}
+
+	g.height = height
+	g.hashAtHeight[height] = *hash
+	g.blockWithHash[*hash] = block
+
+	// Push the new block onto the expiration queue, and remove and old ones if
```
Comment (chappjc) [inline rpcutils/blockgate.go]: and --> any

File: rpcutils/blockgate.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+			delete(g.blockWithHash, oldest.hash)
+		}
+		if !heightInQueue(g.expireQueue, oldest.height) {
+			delete(g.hashAtHeight, oldest.height)
+		}
+	}
+}
+
+func (g *BlockGate) signalHash(blockhash chainhash.Hash) {
+	g.Lock()
```
Comment (chappjc) [inline rpcutils/blockgate.go]: Comment about this lock here and the reason for running this in a deferred goroutine.

File: stakedb/stakedb.go [hunk 11]
```diff
@@ -92,6 +92,8 @@ type StakeDatabase struct {
 	liveTicketCache map[chainhash.Hash]int64
 	poolInfo        *PoolInfoCache
 	PoolDB          *TicketPool
+	waitMtx         sync.Mutex
```
Comment (chappjc) [inline stakedb/stakedb.go]: Make section with comments

File: stakedb/stakedb.go [hunk 12]
```diff
@@ -222,6 +225,50 @@ func NewStakeDatabase(client *rpcclient.Client, params *chaincfg.Params,
...
+func (db *StakeDatabase) LockStakeNode() {
+	db.nodeMtx.RLock()
+}
+
+// UnlockStakeNode unlocks the StakeNode for functions that respect the mutex.
+func (db *StakeDatabase) UnlockStakeNode() {
+	db.nodeMtx.RUnlock()
+}
+
+func (db *StakeDatabase) WaitForHeight(height int64) chan *chainhash.Hash {
```
Comment (chappjc) [inline stakedb/stakedb.go]: Docs

**Comment by chappjc** [comment]
@dcrdata/explorers Requesting a test run by someone else.  No DB rebuilds required.  Should just add robustness to startup.

**Comment by raedah** [comment]
Tested, no probs.

**Pull request merged**